File: plotter.m
clear all;
g = @(n) n; %Predicted g(n) for Insertion
h = @(n) n; %Predicted g(n) for Removing
i = @(n) n; %Predicted g(n) for Worst case lookup
j = @(n) n; %Predicted g(n) for Random lookup
k = @(n) n; %Predicted g(n) for Skewed lookup

orderpreds = {1,2,3,4,5; g, h, i, j, k};

scriptinput = importdata("input.txt");

tests = zeros(scriptinput.data(2)*scriptinput.data(3), 2, 5);
avgs = zeros(scriptinput.data(2), 2, 5);

%Read data
for i = 1:5
   tests(:, :, i) = importdata(strcat("split/t", num2str(i), "split.txt"));
   avgs(:, :, i) = importdata(strcat("splitavg/t", num2str(i), "avg.txt"));
end

%Plot data per subtest with the average of each element amount teste. This result in the time complexity for each subtest.
figure('name',string(scriptinput.textdata))
hold on
for i= 1:5
    plot(avgs(:, 1, i), avgs(:, 2, i)); 
end
title("Complexity of test")
xlabel("Number of elements")
ylabel("Time in ms")
legend("Insert test", "Remove all test", "Non-existent lookup test", "Random lookup test", "Skewed lookup test", 'Location', 'northwest')
hold off


%Plot data per subtest divided with the element amount with the average of each element amount tested. This result in the time complexity for each the relevant table operation for each subtest.
figure('name',string(scriptinput.textdata))
hold on
for i = 1:5
    plot(avgs(:, 1, i), avgs(:, 2, i) ./ avgs(:, 1, i)); 
end
title("Complexity for operations")
xlabel("Number of elements")
ylabel("Time in ms")
legend("Insert test", "Remove all test", "Non-existent lookup test", "Random lookup test", "Skewed lookup test", 'Location', 'northwest')
hold off

%Plot the previous graph divided by a prediction (g(n)) set at the top of the code.
figure('name',string(scriptinput.textdata))
hold on
for i = 1:5
    t = orderpreds(i);
    plot(avgs(:, 1, i), avgs(:, 2, i) ./ avgs(:, 1, i) ./ avgs(:, 1, i)); 
end
title("Prediction of g(n)")
xlabel("Number of elements")
ylabel("Time in ms / g(n)")
legend("Insert test", "Remove all test", "Non-existent lookup test", "Random lookup test", "Skewed lookup test", 'Location', 'northwestoutside')
hold off

%Plots all runs for each subtest in the same graph. Produces a graph similar to the one from the complexity lecture.
figure('name',string(scriptinput.textdata))
hold on
for i = 1:5
    plot(tests(:, 1, i), tests(:, 2, i) ./ tests(:, 1, i), 'x'); 
end
title("Operation complexities for all runs")
xlabel("Number of elements")
ylabel("Time in ms")
legend("Insert test", "Remove all test", "Non-existent lookup test", "Random lookup test", "Skewed lookup test", 'Location', 'northwest')

File: plotter.py
import sys
import matplotlib.pyplot as plt
import numpy as np

testTitles =  ["Insert test", "Remove all test", "Non-existent lookup test", "Random lookup test", "Skewed lookup test"]
class Implementation:

    def __init__(self, title: str, testData, predictions):
        self.predictions = predictions
        self.title = title
        loadedFile = np.loadtxt(testData, delimiter=",")
        self.testData = np.zeros((int(loadedFile.shape[0]/5), 2, 5), np.int32)
        for i in range(5):
            self.testData[:,:,i] = np.delete(loadedFile[np.nonzero(loadedFile[:, 0] == i+1)], 0, 1)
        self.elementAmounts = np.unique(self.testData[:, 0, 0])
        self.avgData = np.zeros((self.elementAmounts.size, 2, 5), np.int32)
        for i in range(5):
            j = 0
            for amount in self.elementAmounts:
                self.avgData[j, 1, i] = np.mean(self.testData[self.testData[:, 0, i] == amount, 1, i])
                self.avgData[j, 0] = amount
                j = j + 1

def plotTimeFunction(implementations: list[Implementation], figureIndex): 
    titles: list[str] = []
    for imp in implementations:
        titles.append(imp.title)

    for i in range(5):
        #plt.subplot(321 + i)
        f = plt.figure(i + figureIndex)
        figureIndex = figureIndex + 1
        plt.title(testTitles[i])
        for imp in implementations:
            plt.plot(imp.avgData[:, 0, i], imp.avgData[:, 1, i] / imp.avgData[:, 0, i])
        plt.legend(titles)
        f.canvas.set_window_title("Time function")

def plotPredictions(implementations: list[Implementation], figureIndex):
    titles: list[str] = []
    for imp in implementations:
        titles.append(imp.title)

    for i in range(5):
        #plt.subplot(321 + i)
        f = plt.figure(i + figureIndex)
        figureIndex = figureIndex + 1
        plt.title(testTitles[i])
        for imp in implementations:
            predictedTimes = np.array([imp.predictions[i](xi) for xi in imp.avgData[:, 0, i]])
            plt.plot(imp.avgData[:, 0, i], (imp.avgData[:, 1, i] / imp.avgData[:, 0, i]) / predictedTimes)
        plt.legend(titles)
        f.canvas.set_window_title("Prediction")

def plotRaw(implementations: list[Implementation], figureIndex):
    titles: list[str] = []
    for imp in implementations:
        titles.append(imp.title)

    for i in range(5):
        #plt.subplot(321 + i)
        f = plt.figure(i + figureIndex)
        figureIndex = figureIndex + 1
        plt.title(testTitles[i])
        for imp in implementations:
            plt.scatter(imp.testData[:, 0, i], imp.testData[:, 1, i] / imp.testData[:, 0, i])
            plt.plot(imp.avgData[:, 0, i], imp.avgData[:, 1, i] / imp.avgData[:, 0, i])
        plt.legend(titles)
        f.canvas.set_window_title("Scatter plot")

def combinePlot(implementations: list[Implementation], figureIndex):
    titles: list[str] = []
    for imp in implementations:
        titles.append(imp.title)

    for imp in implementations:
        #plt.subplot(321 + i)
        f = plt.figure(implementations.index(imp) + figureIndex)
        figureIndex = figureIndex + 1
        plt.title(imp.title)
        for i in range(5):
            plt.plot(imp.avgData[:, 0, i], imp.avgData[:, 1, i] / imp.avgData[:, 0, i])
        plt.legend(testTitles)
        f.canvas.set_window_title("Combined plot")

arr = Implementation("Array table", open(f'array.txt', 'rb'), [lambda x : x, lambda x : x, lambda x : x, lambda x : x, lambda x : x])
mtf = Implementation("MTF table", open(f'mtf.txt', 'rb'), [lambda x : 1, lambda x : x, lambda x : x, lambda x : x, lambda x : x])
dlist = Implementation("Dlist table", open(f'dlist.txt', 'rb'), [lambda x : 1, lambda x : x, lambda x : x, lambda x : x, lambda x : x])

plotTimeFunction([dlist, mtf, arr], 0)
plotPredictions([dlist, mtf, arr], 5)
plotRaw([dlist, mtf, arr], 10)
combinePlot([dlist, mtf, arr], 15)
plt.show()
